mirrord-layer/src/tcp.rs

```diff
@@ -0,0 +1,130 @@
+    //     self.incoming.recv().await
+    // }
+
+    pub async fn listen_request(&self, listen: Listen) -> Result<()> {
+        self.send(TrafficHandlerInput::Listen(listen)).await
+    }
+
+    pub async fn new_tcp_connection(&self, conn: NewTCPConnection) -> Result<()> {
+        self.send(TrafficHandlerInput::NewConnection(conn)).await
+    }
+
+    pub async fn tcp_data(&self, data: TCPData) -> Result<()> {
+        self.send(TrafficHandlerInput::Data(data)).await
+    }
+
+    pub async fn tcp_close(&self, close: TCPClose) -> Result<()> {
+        self.send(TrafficHandlerInput::Close(close)).await
+    }
+}
+
+#[async_trait]
+pub trait TCPHandler {
+    fn ports(&mut self) -> &HashSet<Listen>;
+    fn ports_mut(&mut self) -> &mut HashSet<Listen>;
+
+    /// Returns true to let caller know to keep running
+    async fn handle_incoming_message(&mut self, msg: Option<TrafficHandlerInput>) -> Result<bool>
+    where
+        Self: Send,
+    {
+        if let Some(msg) = msg {
+            match msg {
+                TrafficHandlerInput::NewConnection(conn) => self.handle_new_connection(conn).await,
+                TrafficHandlerInput::Data(data) => self.handle_new_data(data).await,
+                TrafficHandlerInput::Close(close) => self.handle_close(close).await,
+                TrafficHandlerInput::Listen(listen) => self.handle_listen(listen).await,
+            }
+            .unwrap_or_else(|err| error!("message handling failed with {err:?}"));
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Handle NewConnection messages
+    async fn handle_new_connection(&mut self, conn: NewTCPConnection) -> Result<()>;
+
+    /// Connects to the local listening socket, add it to the queue and return the stream.
+    /// Find better name
+    async fn create_local_stream(&mut self, conn: &NewTCPConnection) -> Option<TcpStream> {
+        let listen_data = self.ports().get(&conn.destination_port)?;
+        let addr = match listen_data.ipv6 {
+            false => SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), listen_data.real_port),
+            true => SocketAddr::new(IpAddr::V6(Ipv6Addr::LOCALHOST), listen_data.real_port),
+        };
+
+        let info = SocketInformation::new(SocketAddr::new(conn.address, conn.source_port));
+        {
+            CONNECTION_QUEUE.lock().unwrap().add(&listen_data.fd, info);
+        }
+
+        TcpStream::connect(addr)
+            .await
+            .inspect_err(|err| {
+                error!("create local stream failed, couldn't connect {addr:?} with {err:?}")
+            })
+            .ok()
+    }
+
+    /// Handle New Data messages
+    async fn handle_new_data(&mut self, data: TCPData) -> Result<()>;
+
+    /// Handle connection close
+    async fn handle_close(&mut self, close: TCPClose) -> Result<()>;
+
+    /// Handle listen request
+    async fn handle_listen(&mut self, listen: Listen) -> Result<()> {
```

Comment:
Why return `Result` here if the inner operation can't fail?

Also, do we need this function at all, as it's basically a one-liner?

Reply:
welp the steal might have different logic? also that's why it's Result

Reply:
Hmm, but if they can return in different ways, should they still be treated the same?

If A can never error, but B can, adding a function that displays something like "Hey, A may fail", is not very nice. I prefer that the function signature matches what's inside.

Reply:
What about using Infallible then?

Reply:
Hmm, sure, but I'm still not quite certain that things with such potential different behavior should be shared. Feels a bit to much OOP.